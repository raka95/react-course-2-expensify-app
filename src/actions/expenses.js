import uuid from "uuid";
//Add-Expence

export const addExpense=(
    {
        description='',
        note='',
        amount=0,
        createdAt=0
    }={}
)=>({
    type:"Add-EXPENSE",
    expense:{
        id:uuid(),
        description,
        note,
        amount,
        createdAt
    }
}
);
//Remove-Expense

export const removeExpense=({id}={})=>(
    {
        type:"Remove-Expense",
        id
    }
);
//Edit-Expense

export const editExpense=(id,update)=>({
    type:"Edit-Expense",
    id,
    update
});