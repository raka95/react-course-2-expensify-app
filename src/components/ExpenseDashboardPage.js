// import {Link,NavLink} from  'react-router-dom';
import React from 'react';
import ConnectExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
export const ExpenseDashboardPage=()=>(
    <div>
        <ExpenseListFilters />
        <ConnectExpenseList />
    </div>
);
export default ExpenseDashboardPage;