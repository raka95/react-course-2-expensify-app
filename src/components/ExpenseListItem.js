import React from 'react';
import { NavLink } from 'react-router-dom';


export const ExpenseListItem = ({ description, amount, createdAt, id, dispatch }) => (
    <div>
        <NavLink to={`/edit/${id}`} >
            <h1>{description}</h1>
        </NavLink>
        <p>amount: {amount} - createdAt: {createdAt}</p>
    </div>
);

export default (ExpenseListItem);
// const ExpenseListItem=({description,amount,createdAt})=>(
//     <div>
//         <h1>{description}</h1>
//         <p>amount: {amount} - createdAt: {createdAt}</p>
//     </div>
// );
// export default ExpenseListItem;