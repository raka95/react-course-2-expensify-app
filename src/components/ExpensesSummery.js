import React from 'react';
import {connect} from 'react-redux';
import selectExpenses from "../selectors/expenses";
import selectExpensesTotal from '../selectors/expenses-total';


export const ExpensesSummery=({expense})=>(
    <div>
    <h1>Viewing {expense.length} expenses totalling {selectExpensesTotal(expense)}</h1>
    </div>
);
//{
// //  return expense.reduce((a,b)=>{
// //      return a + b.amount;
// //  },0);
// return expense.map((val)=>val.amount).reduce((sum,value)=>{sum+value,0});
// };

const mapStateToProps=(state)=>({
    expense:selectExpenses(state.expenses,state.filters)
})
export default connect(mapStateToProps)(ExpensesSummery);
