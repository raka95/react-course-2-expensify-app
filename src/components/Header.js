import {Link,NavLink} from  'react-router-dom';
import React from 'react';
const Header=()=>(
    <header>
    <h1>Expensify</h1>
    <NavLink to='/' activeClassName="is-active" exact={true}>Home </NavLink>
    <NavLink to='/create' activeClassName="is-active">Create </NavLink> 
    </header>
);
export default Header;
//   <NavLink to='/help' activeClassName="is-active">Help </NavLink> 