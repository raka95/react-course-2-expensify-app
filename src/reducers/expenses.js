//Expences Reducers  

const expensesReducerDefault=[];
export default (state=expensesReducerDefault,action)=>{
// const expensesReducer=(state=expensesReducerDefault,action)=>{
    switch(action.type){
        case "Add-EXPENSE":
            return [
                ...state,
                action.expense
            ];
        case "Remove-Expense":
            return state.filter(({id})=>id!==action.id);
        case "Edit-Expense":
            return state.map((val)=>{
                if(val.id===action.id)
                    return {
                        ...val,
                        ...action.update
                    };
                    return val;
            })
        default:
            return state;
    }
};
//export default expensesReducer;