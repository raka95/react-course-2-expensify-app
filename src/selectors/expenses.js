// get visible expenses
export default (expenses,{text,sortBy,startDate,endDate})=>{
// const getVisibleExpenses=(expenses,{text,sortBy,startDate,endDate})=>{
    return expenses.filter((val)=>{
     const createdAtMoment=val.createdAt;
     const startDateMAtch=startDate ? startDate.isSameOrBefore(createdAtMoment,"day") :true ;
     const endDateMatch=endDate ?  endDate.isSameOrAfter(createdAtMoment,"day"):true ;
     const textMatch=val.description.toLowerCase().includes(text);
 
     return startDateMAtch&&endDateMatch&&textMatch;
    }).sort((a,b)=>{
        if(sortBy==="date")
         return b.createdAt-a.createdAt;
         else if(sortBy==="amount")
         return b.amount-a.amount;
    })
 };
 