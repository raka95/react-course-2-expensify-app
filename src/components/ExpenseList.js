import React from 'react';
import {connect} from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";


export const ExpenseList=({expense})=>(
    <div>
        {expense.length===0 ? (
            <p>No Expenses</p>
        ):(
            expense.map((val)=>(
                <ExpenseListItem 
                    key={val.id} 
                    {...val}/>
                )
            )
        )}
    </div>
);
//
// 
const mapStateToProps=(state)=> (
    {
        expense:selectExpenses(state.expenses,state.filters)
    }
);
export default connect(mapStateToProps)(ExpenseList);
//

// const ConnectExpenseList=connect((state)=>{
//     return {
//         name:"Ramin",
//         expense:state.expenses,
//         filters:state.filters
//     };
// })(ExpenseList);
// export default ConnectExpenseList;

