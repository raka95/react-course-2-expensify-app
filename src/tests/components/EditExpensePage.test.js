import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import { EditExpensePage } from "../../components/EditExpensePage";
// import { editExpense, removeExpense } from '../../actions/expenses';

let editExpense, removeExpense, history, wrapper;
beforeEach(() => {
    editExpense = jest.fn();
    removeExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(
        <EditExpensePage
            expense={expenses[1]}
            removeExpense={removeExpense}
            editExpense={editExpense}
            history={history}
        />
    );
});

test("Render EditExpensePage to be a class based component", () => {
    expect(wrapper).toMatchSnapshot();
});

test("should handle editExpense", () => {
    wrapper.find("ExpenseForm").prop("onSubmit")(expenses[1]);
    expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
});

test("should handle removeExpense", () => {
    wrapper.find('button').simulate('click');
    expect(removeExpense).toHaveBeenLastCalledWith({ id: expenses[1].id });
    expect(history.push).toHaveBeenLastCalledWith('/');
});
