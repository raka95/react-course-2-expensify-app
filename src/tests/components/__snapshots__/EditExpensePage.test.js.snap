// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render EditExpensePage to be a class based component 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 2000,
        "createdAt": -345600000,
        "description": "rent",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
